src/client/components/immutable-input/immutable-input.tsx

```diff
@@ -49,14 +73,37 @@ export class ImmutableInput extends React.Component<ImmutableInputProps, Immutab
   initFromProps(props: ImmutableInputProps) {
     if (!props.instance || !props.path) return;
 
+    var validString: string;
+
+    if (this.state.validString === undefined) {
+      validString = props.valueToString(ImmutableUtils.getProperty(props.instance, props.path));
+    } else {
+      var currentCanonical = props.valueToString(props.stringToValue(this.state.validString));
+      var possibleCanonical = props.valueToString(ImmutableUtils.getProperty(props.instance, props.path));
+
+      validString = currentCanonical === possibleCanonical ? this.state.validString : possibleCanonical;
+    }
+
     this.setState({
-      newInstance: props.instance,
-      invalidValue: undefined
+      myInstance: props.instance,
+      invalidString: undefined,
+      validString
+    });
+  }
+
+  reset() {
+    this.setState({
+      invalidString: undefined,
+      validString: undefined
     });
   }
 
   componentWillReceiveProps(nextProps: ImmutableInputProps) {
-    if (nextProps.instance !== this.state.newInstance) {
+    if (nextProps.instance === undefined) {
+      this.reset();
+    }
+
+    if (this.state.invalidString === undefined && nextProps.instance !== this.state.myInstance) {
```

Comment:
That wouldn't work. The idea here is to check for strict equality, in case the call to `onChange` failed.